docs/file-formats/kmx-file-format.md

```diff
@@ -0,0 +1,301 @@
+# KMX Binary Format
+
+This document describes the binary format for KMX, compiled from Keyman .kmn
+source files.
+
+## Principles
+
+- All integer values are unsigned 32-bit little-endian unless otherwise
+  specified.
+- All offsets are 32-bit little-endian values. All offsets are relative to the
+  beginning of the data (normally start of file). A zero offset indicates that
+  the value is not present, equivalent to `NULL` -- no offset can ever point to
+  the start of the file.
+- All strings are UTF-16LE unless otherwise specified and null terminated.
+  Strings are referenced by offset (see above).
+- The file format has a header section followed by data. There is no specific
+  defined order for data in the file. Empty space is permitted but should be
+  zeroed. Note that `COMP_KEYBOARD_KMXPLUS` must follow immediately after
+  `COMP_KEYBOARD` with no empty space (that is at byte 0x40).
+
+## `COMP_KEYBOARD`: KMX header
+
+The very first section is a header that describes the remainder of the file.
+
+| ∆ | Bits | Name                 | Description                                                                       |
+|---|------|----------------------|-----------------------------------------------------------------------------------|
+| 0 |  32  | `dwIdentifier`       | `KXTS` 0x5354584B, `kmx.FILEID_COMPILED`                                          |
+| 4 |  32  | `dwFileVersion`      | minimum Keyman version that can read the file, `KMX_Version` enum                 |
+| 8 |  32  | `dwCheckSum`         | deprecated in 16.0, `0x0`; CRC32 checksum of entire file                          |
+|12 |  32  | `KeyboardID`         | deprecated in 10.0, `0x0`; Windows LANGID for keyboard                            |
+|16 |  32  | `IsRegistered`       | deprecated in 10.0, `0x1`; registration status of compiler                        |
+|20 |  32  | `version`            | deprecated, `0x0`; was version of keyboard, see `TSS_KEYBOARDVERSION`             |
+|24 |  32  | `cxStoreArray`       | number of `COMP_STORE` entries in `COMP_KEYBOARD.dpStoreArray`                    |
+|28 |  32  | `cxGroupArray`       | number of `COMP_GROUP` entries in `COMP_KEYBOARD.dpGroupArray`                    |
+|32 |  32  | `dpStoreArray`       | offset of first `COMP_STORE` entry                                                |
+|36 |  32  | `dpGroupArray`       | offset of first `COMP_GROUP` entry                                                |
```

Comment:
I'm confused at the moment.

I've read through this source file, but I'm not clear on how we know the actual full _byte length_ of many pieces of the data indirectly referenced by, but ultimately represented by, `dpStoreArray` and `dpGroupArray`. 

----

For `dpStoreArray`:

> ```
> | 4 |  32  | `dpName`     | offset to name of the store from keyboard source, `0x0` if not present |
> | 8 |  32  | `dpString`   | offset to value of the store     
> ```

So... those just store offsets; the data isn't inlined.  We do get a fixed-length `COMP_STORE` "subheader" / table definition, so to speak, so I'm fine parsing this section... just not the data it references. To me, this just passes the buck; how do we know the length of the encoded values for each store?  Of the data pointed to by each `dpName` and `dpString` entry per store?
- Is it implicit in some form?

My current best guess:
- Suppose two stores have been defined:  `a` and `b`.
- We parse `a`'s header.
    - We see the `dpName` (for `a`)'s offset.  We don't know its length yet, but `dpString` is coming soon, so we wait.
    - `dpString` (for `a`) - `dpName` (for `a`) = `dpName` (for `a`)'s length, implicitly defined - we now know this, so we can retrieve it
    - `dpName` - we know its _start_ location, but not its length... yet.  We "note" somehow that the next thing to process is this value and wait for the next _overall_ header entry with an offset index.
- We now parse `b`'s header.
    - Upon seeing `dpName`'s offset for `b`, we now have a new offset.
        - `dpName` (for `b`) - `dpString` (for `a`) yields `dpString` (for `a`)'s length; we can now finish parsing data for store `a`.

Of course, we'd eventually reach the end of processing `COMP_STORE`... with the last such `dpString` still awaiting the next offset entry.  Which would come from parsing `COMP_GROUP`.

This feels wrong due to being... maybe not too complex, but too unencapsulated.  Too messy.  _Especially_ since this guess involves delaying processing of a subsection, sometimes until we receive the relevant offset from a completely separate, higher-level section.  So I'm thinking - or at least, _hoping_ - that my guess here is wrong.  But this is the best answer I can infer from what I see in the documentation at present.

And again, this guess is based _solely_ on what I see in the documentation, as someone who's neither so much as looked into the KMX format before nor inspected the binary data stored within any KMX files.

I have the same issue with `dpGroupArray` - it looks like the same patterns result, especially once down to the `COMP_KEY` level.

So far, I know how to parse the headers, but I don't know how to properly subdivide the rest of the file to safely parse and load the store and rule _definitions_.  (The declarations are mostly cool; I just don't know the lengths at the lower levels.)  If I don't know how to determine the length of ranges being pointed to, I'm out of luck if working from this documentation alone.

Reply:
Apologies for heavy comment-editing here.  I'm fairly certain that we're relying on string null-termination now, but failed to originally connect those dots before.

My old guess, before I gear-shifted to thinking based fully on null-termination, would use the following pattern, hidden within a `<details>` below since it's probably wrong.

<details><summary>A pattern I hope is wrong</summary>

- Suppose two stores have been defined:  `a` and `b`.
- We parse `a`'s header.
    - We see the `dpName` (for `a`)'s offset.  We don't know its length yet, but `dpString` is coming soon, so we wait.
    - `dpString` (for `a`) - `dpName` (for `a`) = `dpName` (for `a`)'s length, implicitly defined - we now know this, so we can retrieve it
    - `dpName` - we know its _start_ location, but not its length... yet.  We "note" somehow that the next thing to process is this value and wait for the next _overall_ header entry with an offset index.
- We now parse `b`'s header.
    - Upon seeing `dpName`'s offset for `b`, we now have a new offset.
        - `dpName` (for `b`) - `dpString` (for `a`) yields `dpString` (for `a`)'s length; we can now finish parsing data for store `a`.

Of course, we'd eventually reach the end of processing `COMP_STORE`... with the last such `dpString` still unable to parse due to awaiting the next offset entry... which would come from parsing `COMP_GROUP`.  🤮 

Then again... right, C++, so we may be null-terminating the string.  We could rely on that... though this means that the KMX parser has no real way to skip sections, as it has to parse each subsection 

This feels wrong due to being... maybe not too complex, but too unencapsulated.  Too messy.  _Especially_ since this guess involves delaying processing of a subsection, sometimes until we receive the relevant offset from a completely separate, higher-level section.  So I'm thinking - or at least, _hoping_ - that my guess here is wrong.  But this is the best answer I can infer from what I see in the documentation at present.

And again, this guess is based _solely_ on what I see in the documentation, as someone who's neither so much as looked into the KMX format before nor inspected the binary data stored within any KMX files.

I have the same issue with `dpGroupArray` - it looks like the same patterns result, especially once down to the `COMP_KEY` level.

So far, I know how to parse the headers, but I don't know how to properly subdivide the rest of the file to safely parse and load the store and rule _definitions_.  (The declarations are mostly cool; I just don't know the lengths at the lower levels.)  If I don't know how to determine the length of ranges being pointed to, I'm out of luck if working from this documentation alone.

</details>

Reply:
Note the timeline of this comment chain; I've spent over an hour iterating over this to improve my understanding and reach the current point displayed above, and that's with prior knowledge of Keyman keyboard rules, stores, etc.

Reply:
You need to read the rest of the document with reference to the principles at the top of the document:

```
- All offsets are 32-bit little-endian values. All offsets are relative to the
  beginning of the data (normally start of file). A zero offset indicates that
  the value is not present, equivalent to `NULL` -- no offset can ever point to
  the start of the file.
- All strings are UTF-16LE unless otherwise specified and null terminated.
  Strings are referenced by offset (see above).
```

I think you are overthinking how the file is loaded. It's essentially a 32-bit memory mapped file. In a 32-bit process, we load the file into a memory buffer, then rewrite all the `dp*` members to add the base of the buffer. Done. No need to scan strings, etc, they are already in C++ zero terminated format. The structures are fixed sizes. (For 64-bit there is a fixup we have to do, but the format was invented before 64-bit was a thing -- at least on Windows.)

I'll add a note about memory mapping in the Principles section.

> ... "jumbled mess" ...

> I can see ways within either scenario in which we could insert a practically-random payload in the middle of the KMX file somewhere such that the parser would essentially ignore it.

Yes. Again, refer to the principles:

```
- The file format has a header section followed by data. There is no specific
  defined order for data in the file. Empty space is permitted but should be
  zeroed. Note that `COMP_KEYBOARD_KMXPLUS` must follow immediately after
  `COMP_KEYBOARD` with no empty space (that is at byte 0x40).
```

Reply:
Ah, must have missed that.  Admittedly, when I started my review... I accidentally did so from the view of just the last commit.  Thought I went back and covered things, but in retrospect I didn't double-check the very top.

Glad to see that it is explained; that addresses my concerns, then.